File: java/PlanesAndroid/app/src/main/java/com/planes/common/VectorIterator.java
package com.planes.common;

import java.util.Vector;

public class VectorIterator<T> {

    public VectorIterator() {
        m_internalList = new Vector<T>();
        reset();
    }

    public void reset() {
        m_idx = -1;
    }

    public boolean hasNext() {
        return (m_idx < m_internalList.size() - 1);
    }

    public T next() {
        m_idx++;
        return m_internalList.get(m_idx);
    }

    public int itemNo() {
        return m_internalList.size();
    }

    //for test purposes
    public void setInternalList(final Vector<T> list) {
        m_internalList = list;
    }

    protected Vector<T> m_internalList;
    private int m_idx;
}

File: kotlin/PlanesAndroid/app/src/main/java/com/planes_multiplayer/single_player_engine/VectorIterator.kt
package com.planes_multiplayer.single_player_engine

import java.util.*

open class VectorIterator<T> {
    protected var m_internalList: Vector<T>
    private var m_idx = 0

    init {
        m_internalList = Vector()
        reset()
    }

    fun reset() {
        m_idx = -1
    }

    operator fun hasNext(): Boolean {
        return m_idx < m_internalList.size - 1
    }

    operator fun next(): T {
        m_idx++
        return m_internalList[m_idx]
    }

    fun itemNo(): Int {
        return m_internalList.size
    }

    //for test purposes
    fun setInternalList(list: Vector<T>) {
        m_internalList = list
    }

}